src/aws-manager.js

```diff
@@ -139,10 +125,221 @@ class AwsManager {
+    });
+
+    let data = JSON.stringify({
+      managedSpotRequests: this.__spotRequestIdCache || [],
+      internallyTrackedSpotRequests: this.__internalState || [],
+    }, null, 2);
+
+    //fs.writeFileSync('internal-data.json', data);
+    return this.spotRequestContainer.write('internal-provisioner-data', data);
+  }
+
+  /**
+   * Given a resource (Instance or SpotInstanceRequest ec2 data type) and a
+   * region, determine using a variety of means to see if we can map the
+   * resource to a WorkerType.
+   */
+  async workerTypeForResource(resource, region) {
+    assert(typeof resource === 'object');
+    assert(typeof region === 'string');
+
+    let srid = resource.SpotInstanceRequestId;
+    let instanceId = resource.InstanceId;
+
+    // First, we should check to see if we have the mapping of
+    // SpotInstanceRequestId in memory (which we persist to an azure blob
+    // storage blob).  If this item is in that cache, we can be sure that the
+    // provisioner both owns it because it created it, and also which worker
+    // type, since we add that information to the cache when we add the id.
+    let found = this.__spotRequestIdCache.filter(x => x.id === srid && x.region === region);
+    if (found.length === 1) {
+      let workerType = found[0].workerType;
+      log.info({workerType}, 'found workerType in spotRequestIdCache');
+      return workerType;
+    } else if (found.length > 1) {
+      let firstX = found[0].workerType;
+      // making a wasted comparison of the first is probably faster than
+      // slicing the array
+      for (let x of found) {
+        if (firstX !== x.workerType) {
+          let err = new Error('Multiple worker type possiblities found');
+          log.error({err, workerTypesFound: found.map(x => x.workerType)}, 'found too many');
+          throw err;
+        }
+      }
+    }
+
+    // Next, let's test if the instance has already had tags applied.  We
+    // cannot rely on tags being there because there is a two-step process for
+    // tagging, however, we can assume that if an instance has the right tag
+    // that it's correct and is how we should categorise things.  Since we only
+    // tag on the following iteration, we might get into the situation where a
+    // resource is created, forgotten about, then untagged.  For these cases,
+    // we'll continue later on and use the instanceId checks when the request
+    // turns into an instance.
+    if (resource.TagSet) {
+      let owned = false;
+      let workerType;
+      for (let tag of resource.TagSet) {
+        if (owned && workerType) {
+          break; // we already have what we need
+        }
+        switch (tag.Key) {
+          case 'Owner':
+            if (tag.Value === this.provisionerId) {
+              owned = true;
+            }
+            break;
+          case 'Name':
+            workerType = tag.Value;
+            break;
+          default:
+            break;
+        }
+      }
```

Comment:
You really should start to use lodash for things like this.... This would be:
```js
let workerType = (_.find(resource.TagSet, {Key: 'Name'}) || {}).Value;
```

Really, that's it... Less code = less bugs. And it's a lot faster to read...
(sure, you have to look up the lodash docs to check that _.find() will work like this. But when reading the code the intend is much clearer without much analysis.